cfy_manager/components/cluster/cluster.py

```diff
@@ -0,0 +1,187 @@
+#########
+# Copyright (c) 2018 Cloudify Platform Ltd. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+#  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  * See the License for the specific language governing permissions and
+#  * limitations under the License.
+import os
+import json
+import random
+import string
+import requests
+import tempfile
+
+from ...components.base_component import BaseComponent
+from ...utils import common
+from ...logger import get_logger
+from ...exceptions import BootstrapError, NetworkError
+from ...config import config
+from ...components.service_names import (
+    CLUSTER,
+    MANAGER,
+    POSTGRESQL_CLIENT
+)
+from ...components.components_constants import (
+    SERVICES_TO_INSTALL,
+    MASTER_IP,
+    NODE_NAME,
+    PUBLIC_IP,
+    PREMIUM_EDITION
+)
+from ...constants import REST_HOME_DIR
+from ...components.service_components import DATABASE_SERVICE
+from ...utils.network import get_auth_headers
+
+CLUSTER_REMOVE_SCRIPT = '/opt/manager/env/bin/teardown_cluster'
+CLUSTER_SYSTEMD_UNIT_NAME = 'ha-cluster'
+NODE_NAME_GENERATED_CHAR_SIZE = 6
+
+logger = get_logger('cluster')
+
+
+class ClusterComponent(BaseComponent):
+    def __init__(self, skip_installation):
+        super(ClusterComponent, self).__init__(skip_installation)
+
+    def _generic_cloudify_rest_request(self, host, port, path,
+                                       method, data=None):
+        url = 'http://{0}:{1}/api/{2}'.format(host, port, path)
+        try:
+            if method == 'get':
+                response = requests.get(url, headers=get_auth_headers())
+            elif method == 'put':
+                response = requests.put(url, json=data,
+                                        headers=get_auth_headers())
+            else:
+                response = None
+        # keep an erroneous HTTP response to examine its status code, but still
+        # abort on fatal errors like being unable to connect at all
+        except requests.HTTPError as e:
+            response = e
+        except requests.URLRequired as e:
+            raise NetworkError(
+                'REST service returned an invalid response: {0}'.format(e))
+        if response.status_code == 401:
+            raise NetworkError(
+                'Could not connect to the REST service: '
+                '401 unauthorized. Possible access control misconfiguration'
```

Comment:
this requires the same credentials for the master and for the replica, right? possibly note that in the error message

Reply:
Fixed